src/WinRT.Runtime2/InteropServices/RestrictedErrorInfo.cs

```diff
@@ -28,10 +34,127 @@ public static unsafe class RestrictedErrorInfo
     /// calls (both in managed and native code). This improves the debugging experience.
     /// </remarks>
     /// <seealso cref="Marshal.GetExceptionForHR(int)"/>
-    public static Exception? GetExceptionForHR(HRESULT errorCode)
+    public static Exception GetExceptionForHR(HRESULT errorCode, out bool restoredExceptionFromGlobalState)
     {
-        // TODO
-        return null;
+        restoredExceptionFromGlobalState = false;
+        Exception? ex;
+        string errorMessage = string.Empty;
+        Exception internalGetGlobalErrorStateException;
+
+        WindowsRuntimeObjectReferenceValue restrictedErrorInfoValue = ExceptionHelpers.BorrowRestrictedErrorInfo();
+
+        try
+        {
+            void* restrictedErrorInfoValuePtr = restrictedErrorInfoValue.GetThisPtrUnsafe();
+            if (restrictedErrorInfoValuePtr != default)
+            {
+                if (Marshal.QueryInterface((IntPtr)restrictedErrorInfoValuePtr, WellKnownInterfaceIds.IID_ILanguageExceptionErrorInfo, out nint languageErrorInfoPtr) >= 0)
```

Comment:
For consistency, use `IUnknownVftbl.QueryInterfaceUnsafe` instead of `Marshal.QueryInterface`